_source/_posts/2022-08-31-state-of-java-project-loom.md

```diff
@@ -0,0 +1,263 @@
+
+# Average time
+Benchmark                             Mode  Cnt  Score   Error  Units
+LoomBenchmark.platformThreadPerTask   avgt    5  5.600 ± 0.768   s/op
+LoomBenchmark.platformThreadPool      avgt    5  3.887 ± 0.717   s/op
+LoomBenchmark.virtualThreadPerTask    avgt    5  1.098 ± 0.020   s/op
+```
+
+You can find the benchmark [source code on GitHub](https://github.com/deepu105/java-loom-benchmarks). Here are some other meaningful benchmarks for virtual threads:
+
+- An interesting benchmark using ApacheBench [on GitHub](https://github.com/ebarlas/project-loom-comparison) by [Elliot Barlas](https://twitter.com/ElliotBarlas)
+- A benchmark using Akka actors [on Medium](https://medium.com/@zakgof/a-simple-benchmark-for-jdk-project-looms-virtual-threads-4f43ef8aeb1) by [Alexander Zakusylo](https://medium.com/@zakgof)
+- JMH benchmarks for I/O and non-I/O tasks [on GitHub](https://github.com/colincachia/loom-benchmark) by [Colin Cachia](https://twitter.com/colincachia)
+
+### Structured Concurrency
+
+> Structured concurrency will be an incubator feature in Java 19.
+
+[Structured Concurrency](https://openjdk.org/jeps/428) aims to simplify multi-threaded and parallel programming. It treats multiple tasks running in different threads as a single unit of work, streamlining error handling and cancellation while improving reliability and observability. This helps to avoid issues like thread leaking and cancellation delays. Being an incubator feature, this might go through further changes during stabilization.
+
+Consider the following example using `java.util.concurrent.ExecutorService`.
+
+```java
+void handleOrder() throws ExecutionException, InterruptedException {
+    try (var esvc = new ScheduledThreadPoolExecutor(8)) {
+        Future<Integer> inventory = esvc.submit(() -> updateInventory());
+        Future<Integer> order = esvc.submit(() -> updateOrder());
+
+        int theInventory = inventory.get();   // Join updateInventory
+        int theOrder = order.get();           // Join updateOrder
+
+        System.out.println("Inventory " + theInventory + " updated for order " + theOrder);
+    }
+}
+```
+
+We want `updateInventory()` and `updateOrder()` subtasks to be executed concurrently. Each of those can succeed or fail independently. Ideally, the `handleOrder()` method should fail if any subtask fails. However, if a failure occurs in one subtask, things get messy.
+
+- Imagine that `updateInventory()` failed and throws an exception. Then, the `handleOrder()` method will throw an exception when calling `inventory.get()`. So far is fine, but what about `updateOrder()`? Since it runs on its own thread, it can complete successfully. But now we have an issue with a mismatch in inventory and order. Suppose the `updateOrder()` is an expensive operation. In that case, we are just wasting the resources for nothing, and we would have to write some sort of guard logic to revert the updates done to order as our overall operation failed.
```

Comment:
```suggestion
- Imagine that `updateInventory()` fails and throws an exception. Then, the `handleOrder()` method throws an exception when calling `inventory.get()`. So far this is fine, but what about `updateOrder()`? Since it runs on its own thread, it can complete successfully. But now we have an issue with a mismatch in inventory and order. Suppose the `updateOrder()` is an expensive operation. In that case, we are just wasting the resources for nothing, and we will have to write some sort of guard logic to revert the updates done to order as our overall operation has failed.
```
tense alignment